impl/src/into_iterator.rs

```diff
@@ -26,16 +26,19 @@ pub fn expand(input: &DeriveInput, trait_name: &'static str) -> Result<TokenStre
         let reference_with_lifetime = ref_type.reference_with_lifetime();
 
         let generics_impl;
-        let generics =
-            add_extra_ty_param_bound_ref(&input.generics, trait_path, ref_type);
-        let (_, ty_generics, where_clause) = generics.split_for_impl();
         let (impl_generics, _, _) = if ref_type.is_ref() {
-            generics_impl = add_extra_generic_param(&generics, lifetime.clone());
+            generics_impl = add_extra_generic_param(&input.generics, lifetime.clone());
             generics_impl.split_for_impl()
         } else {
-            generics.split_for_impl()
+            input.generics.split_for_impl()
         };
-        // let generics = add_extra_ty_param_bound(&input.generics, trait_path);
+
+        let generics = add_extra_where_clauses(
+            &input.generics,
+            quote! { where #reference_with_lifetime #field_type: #trait_path },
```

Comment:
@MegaBluejay how about reworking the whole `derive(IntoIterator)` macro implementation in the style the `derive(Into)` macro is implemented: defined typed attribute and expansion structure?

It also would be nice to add `tests/compile_fail/into_iter`.

And we may rename the whole PR as "`IntoIterator` macro rework" or something.

Reply:
Personally, to keep reviewing easy, I prefer to keep refactoring separate from bugfixes/features, if possible. PRs that do refactoring are usually hard to review, because they change lots of code. Reviewing is made a lot easier if you can assume that the changed code should do the same. Once you also start adding features/bugfixes to such a changeset it becomes really hard to understand if a change in behaviour was unintended, or if it was necessary to for the feature/bugfix.

Adding tests that are currently missing always appreciated though. I don't think it's necessary to block the PR on that though. The change seems simple enough, that I wouldn't expect it to introduce breakage.

Reply:
@MegaBluejay well then... let's make a rework in a separate PR. For now, only add a rich set of tests.